services/venice-controller/src/main/java/com/linkedin/venice/controller/kafka/protocol/serializer/NewSemanticUsageValidator.java

```diff
@@ -0,0 +1,253 @@
+    }
+
+    Schema subSchema = findObjectSchemaInsideUnion(object, currentField);
+    if (subSchema == null) {
+      return returnTrueAndLogError(
+          String.format(
+              "Field %s: Cannot find the match schema for value %s from schema union",
+              formatFieldName(currentField.name()),
+              object));
+    }
+
+    Schema.Field newCurrentField = AvroCompatibilityHelper.createSchemaField(currentField.name(), subSchema, "", null);
+    return isNonDefaultValueField(object, newCurrentField);
+  }
+
+  /**
+   * Find the schema for the object inside the union.
+   * @param currentField the field with union type
+   * @param object the object to find the schema
+   * @return the schema for the object if found, null otherwise
+   */
+  public static Schema findObjectSchemaInsideUnion(Object object, Schema.Field currentField) {
+    for (Schema subSchema: currentField.schema().getTypes()) {
+      if (isCorrectSchema(object, subSchema)) {
+        return subSchema;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Check if the value is non-default for the given field.
+   * If the value is null, it is considered default.
+   * Default value of each type are only applicable for non-nested fields.
+   * If the field is nested-field, the default value is always null.
+   *
+   * @param object: the value to check
+   * @param field: the field to check
+   * @return true if the value is non-default, false otherwise
+   */
+  public boolean isNonDefaultValueField(Object object, Schema.Field field) {
+    if (object == null)
+      return false;
+
+    Object fieldDefaultValue = AvroSchemaUtils.getFieldDefault(field);
+    Object typeDefaultValue = getDefaultForType(field.schema().getType());
+
+    if (!Objects.equals(object, typeDefaultValue) && !Objects.equals(object, fieldDefaultValue)) {
+      return returnTrueAndLogError(
+          String.format(
+              "Field: %s contains non-default value. Actual value: %s. Default value: %s or %s",
+              formatFieldName(field.name()),
+              object,
+              fieldDefaultValue,
+              typeDefaultValue));
+    }
+    return false;
+  }
+
+  /**
+   * Return true and log the error message.
+   * @param message the error message
+   */
+  private boolean returnTrueAndLogError(String message) {
+    errorMessage.set(message);
+    return true;
+  }
+
+  /**
+   * Format the field name to replace "_" with ".". Reason is that we use "_" as a delimiter in the field name since "." is
+   * invalid character in field name.
+   */
+  private static String formatFieldName(String fieldName) {
+    return fieldName.replace("_", ".");
+  }
+
+  /**
+   * Get the default value for the given type.
+   * @param type the type
+   * @return the default value if it is primitive type, else null for nested fields.
+   */
+  private static Object getDefaultForType(Schema.Type type) {
+    switch (type) {
+      case STRING:
+        return "";
```

Comment:
I believe default for String in java would be null.

Reply:
If the object is null, it should not go into this far. So putting the default value for string as null doesn't matter much I think.